Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h

```diff
@@ -49,30 +50,61 @@ ALWAYS_INLINE bool throw_dom_exception(JS::VM& vm, JS::GlobalObject& global_obje
     return false;
 }
 
-template<typename F, typename T = decltype(declval<F>()()), typename Ret = Conditional<!IsExceptionOr<T> && !IsVoid<T>, T, JS::Value>>
-Ret throw_dom_exception_if_needed(auto&& vm, auto&& global_object, F&& fn)
+namespace Detail {
+
+template<typename T>
+struct ExtractExceptionOrValueType {
+    using Type = T;
+};
+
+template<typename T>
+struct ExtractExceptionOrValueType<DOM::ExceptionOr<T>> {
+    using Type = T;
+};
+
+template<>
+struct ExtractExceptionOrValueType<void> {
+    using Type = JS::Value;
+};
+
+template<>
+struct ExtractExceptionOrValueType<DOM::ExceptionOr<void>> {
+    using Type = JS::Value;
+};
+
+}
+
+template<typename T>
+using ExtractExceptionOrValueType = typename Detail::ExtractExceptionOrValueType<T>::Type;
+
+// Return type depends on the return type of 'fn' (when invoked with no args):
+// void or ExceptionOr<void>: Optional<JS::Value>, always returned as missing
+// ExceptionOr<T>: Optional<T>
+// T: Optional<T>
+template<typename F, typename T = decltype(declval<F>()()), typename Ret = Conditional<!IsExceptionOr<T> && !IsVoid<T>, T, ExtractExceptionOrValueType<T>>>
+Optional<Ret> throw_dom_exception_if_needed(auto&& vm, auto&& global_object, F&& fn)
 {
     if constexpr (IsExceptionOr<T>) {
         auto&& result = fn();
         if (throw_dom_exception(vm, global_object, result))
-            return JS::Value();
+            return {};
         if constexpr (requires(T v) { v.value(); })
             return result.value();
-        return JS::Value();
+        return {};
     } else if constexpr (IsVoid<T>) {
         fn();
-        return JS::js_undefined();
+        return {};
     } else {
         return fn();
     }
 }
 
 template<typename T>
-bool should_return_empty(T&& value)
+bool should_return_empty(const Optional<T>& value)
 {
     if constexpr (IsSame<JS::Value, T>)
-        return value.is_empty();
-    return false;
+        return !value.has_value() || value.value().is_empty();
+    return !value.has_value();
```

Comment:
I'm not sure if this is 100% right, we're gonna be returning empty values in these cases:
- The operation threw an exception
- The operation returned void

I'm not sure if we're supposed to be returning `undefined` for the second case, or an empty value?

(cc @linusg)

Reply:
> The operation threw an exception

Empty value is fine.

> The operation returned void

Needs to be undefined I think, as `undefined foo();` in IDL maps to `void foo();` in C++.

Reply:
Alright, I'll make it do that then : )